fix: make app tabs scrollable

BrunoQuaresma opened the pull request:
The solution follows the way VSCode manages its tabs. Since many other editors use a similar approach, this feels familiar and intuitive for our users.

**Demo:**

https://github.com/user-attachments/assets/b4cfb307-268b-4c8b-ac7f-d01dff4ce60b

Fix https://github.com/coder/coder/issues/19438

BrunoQuaresma committed:
fix: move overflowing apps into a dropdown

BrunoQuaresma committed:
use horizontal scrolling

BrunoQuaresma committed:
remove unecessary attr

aslilac commented on site/src/utils/workspace.tsx:
```diff
@@ -286,6 +286,12 @@ export const paramsUsedToCreateWorkspace = (
 
+export const getAgents = (workspace: TypesGen.Workspace) => {
+	return workspace.latest_build.resources
+		.flatMap((r) => r.agents)
+		.filter((a) => !!a);
```
- explicit boolean cast is clearer than `!!` imo
- doing the `filter` first is more efficient, prevents allocating array space for `undefined` entries

```suggestion
		.filter(Boolean)
		.flatMap((r) => r.agents);
```

BrunoQuaresma replied:
- `r.agents` can return `undefined`, which is why the filter happens afterward.
- Instead of using `Boolean`, maybe `a !== undefined` would be clearer in this case. What do you think?

aslilac replied:
also I'm now realizing this is another thing getting moved to utils/ that could be better off somewhere else. maybe modules/workspaces/agents.ts?

BrunoQuaresma replied:
I made a small change to the code so there are no new additions to the workspace utils haha.

aslilac commented on site/src/pages/TaskPage/TaskApps.tsx:
```diff
@@ -17,74 +18,68 @@ import { type FC, useState } from "react";
-	const [activeAppId, setActiveAppId] = useState<string | undefined>(
-		embeddedApps[0]?.app.id,
-	);
+	const embedApps = apps.filter((app) => !app.external);
+	const externalApps = apps.filter((app) => app.external);
```
I know it was already like this, but fact that these groups are perfect partitions makes iterating over the list twice feel a little bad. I'd probably do something like

```
const embeddedApps = [];
const externalApps = [];

for (const app of apps) {
  if (app.external) {
    externalApps.push(app);
  } else {
    embeddedApps.push(app);  
  }
}

BrunoQuaresma committed:
apply PR review suggestions

BrunoQuaresma committed:
Merge branch 'main' of https://github.com/coder/coder into bq/fix-apps-overflowing

BrunoQuaresma committed:
fixes

BrunoQuaresma commented:
@aslilac I’m going to re-request your review since I had to refactor a bit to fix conflicts with main. The changes aren’t too significant, though.

BrunoQuaresma commented:
@kacpersaw I’m requesting your review since you added the subdomain warning in the task feature and I had to resolve a few conflicts. The biggest changes are that I moved it into the app iframe component (so each app can handle its own state) and I added a story to ensure it’s tested. Please let me know if you have any questions!

BrunoQuaresma committed:
fix and add new stories

aslilac commented on site/src/pages/TaskPage/TaskAppIframe.tsx:
```diff
@@ -7,55 +6,43 @@ import {
-		.find((a) => a.apps.some((a) => a.id === app.id));
-
-	if (!agent) {
-		throw new Error(`Agent for app ${app.id} not found in task workspace`);
```
ty for getting rid of this `throw` btw

aslilac commented on site/src/pages/TaskPage/TaskAppIframe.tsx:
```diff
@@ -7,55 +6,43 @@ import {
 		workspace: task.workspace,
 	});
-
-	const appHref = (): string => {
```
ooooh I see, it used to be here. still, I'm really not convinced that this actually does anything, and if it does do anything, it needs a comment documenting what that is in the new location.

BrunoQuaresma commented on site/src/modules/tasks/tasks.ts:
```diff
@@ -1,8 +1,35 @@
+export function getTaskApps(task: Task): WorkspaceAppWithAgent[] {
+	return (
+		task.workspace.latest_build.resources
+			.flatMap((r) => r.agents)
+			.filter((a) => a !== undefined)
```
Sorry about that—I’ll fix it right away.

BrunoQuaresma committed:
apply pr review comments

BrunoQuaresma committed:
fix lint

aslilac commented on site/src/pages/TaskPage/TaskApps.tsx:
```diff
@@ -9,104 +8,67 @@ import {
-		activeApp?.subdomain && !proxy.proxy?.wildcard_hostname;
+	const apps = getTaskApps(task);
+	const [embeddedApps, externalApps] = splitEmbeddedAndExternalApps(apps);
+	const [activeAppId, setActiveAppId] = useState(embeddedApps.at(0)?.id);
```
```suggestion
	const [activeAppId, setActiveAppId] = useState(embeddedApps[0]?.id);
```

BrunoQuaresma commented:
@aslilac thanks a lot for all your reviews! I’m going to merge this PR now, but please let me know if I missed anything.
